src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,347 @@
+            .await
+            .unwrap();
+
+        let mut time_lease = None;
+        while let Some(item) = stream.next().await {
+            match item {
+                Ok(ReplicationMessage::XLogData(xlog_data)) => {
+                    use LogicalReplicationMessage::*;
+
+                    match xlog_data.data() {
+                        Begin(_) => {
+                            // Lease a tx time
+                            time_lease = Some(timestamper.lease().await);
+                        }
+                        Insert(insert) if insert.rel_id() == rel_id => {
+                            let time = time_lease
+                                .as_deref()
+                                .cloned()
+                                .expect("got row event without a prior BEGIN event");
+                            let row = self.row_from_tuple(insert.tuple(), &col_types);
+                            co.yield_(Ok((row, time, 1))).await;
+                        }
+                        Update(update) if update.rel_id() == rel_id => {
+                            let time = time_lease
+                                .as_deref()
+                                .cloned()
+                                .expect("got row event without a prior BEGIN event");
+                            let old_row =
+                                self.row_from_tuple(update.old_tuple().unwrap(), &col_types);
+                            co.yield_(Ok((old_row, time, -1))).await;
+
+                            let new_row = self.row_from_tuple(update.new_tuple(), &col_types);
+                            co.yield_(Ok((new_row, time, 1))).await;
+                        }
+                        Delete(delete) if delete.rel_id() == rel_id => {
+                            let time = time_lease
+                                .as_deref()
+                                .cloned()
+                                .expect("got row event without a prior BEGIN event");
+                            let row = self.row_from_tuple(delete.old_tuple().unwrap(), &col_types);
+                            co.yield_(Ok((row, time, -1))).await;
+                        }
+                        Commit(commit) => {
+                            cur_lsn = commit.commit_lsn().into();
+                            // Release the lease
+                            time_lease = None;
+                        }
+                        Origin(_) | Relation(_) | Type(_) => (),
+                        // TODO turn this into a stream error
+                        Truncate(_) => panic!("source table got truncated"),
+                        // The enum is marked as non_exaustive. Better to be conservative here in
+                        // case a new message is relevant to the semantics of our source
+                        // TODO: turn into a stream error
+                        _ => panic!("unexpected logical replication message"),
```

Comment:
I am hitting this when an ```INSERT``` is performed on a table with a PK. Unfortnuately, I could not print out the offending message.

```2021-02-22T12:57:06.050819Z ERROR panic: unexpected logical replication message
thread: timely:work-0
location: src/dataflow/src/source/postgres.rs:309:30
version: 0.7.1-dev (5e4354b73ec779336995b81ba6f44f16616460bb)
backtrace:
   0: materialized::handle_panic
             at var/lib/buildkite-agent/builds/buildkite-builders-i-099ee983accf64f75-1/materialize/tests/src/materialized/src/bin/materialized/main.rs:618:21
   1: core::ops::function::Fn::call
             at usr/local/lib/rustlib/src/rust/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:597:17
   3: std::panicking::begin_panic::{{closure}}
             at usr/local/lib/rustlib/src/rust/library/std/src/panicking.rs:522:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at usr/local/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:141:18
   5: std::panicking::begin_panic
             at usr/local/lib/rustlib/src/rust/library/std/src/panicking.rs:521:12
   6: dataflow::source::postgres::PostgresSimpleSource::into_generator::{{closure}}
             at var/lib/buildkite-agent/builds/buildkite-builders-i-099ee983accf64f75-1/materialize/tests/src/dataflow/src/source/postgres.rs:309:30
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at usr/local/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
      <genawaiter::core::Advance<F,A> as core::future::future::Future>::poll
             at cargo/registry/src/github.com-1ecc6299db9ec823/genawaiter-0.99.1/src/core.rs:86:15
      genawaiter::sync::stream::<impl futures_core::stream::Stream for genawaiter::sync::generator::Gen<Y,(),F>>::poll_next
             at cargo/registry/src/github.com-1ecc6299db9ec823/genawaiter-0.99.1/src/sync/stream.rs:17:15
   8: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.12/src/stream.rs:129:9
      <futures_util::stream::stream::map::Map<St,F> as futures_core::stream::Stream>::poll_next
             at cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.12/src/stream/stream/map.rs:61:26
   9: <futures_util::stream::stream::fuse::Fuse<S> as futures_core::stream::Stream>::poll_next
             at cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.12/src/stream/stream/fuse.rs:56:27
      futures_util::stream::select::poll_inner
             at cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.12/src/stream/select.rs:112:24
      <futures_util::stream::select::Select<St1,St2> as futures_core::stream::Stream>::poll_next
             at cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.12/src/stream/select.rs:97:13
  10: <I as timely::dataflow::operators::to_stream::ToStreamAsync<T,D>>::to_stream::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/dataflow/operators/to_stream.rs:206:51
      timely::dataflow::operators::generic::operator::source::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/dataflow/operators/generic/operator.rs:576:13
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/dataflow/operators/generic/builder_rc.rs:163:26
      <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/dataflow/operators/generic/builder_raw.rs:203:9
  11: timely::progress::subgraph::PerOperatorState<T>::schedule
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/progress/subgraph.rs:656:30
  12: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/progress/subgraph.rs:338:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/progress/subgraph.rs:304:17
  13: timely::progress::subgraph::PerOperatorState<T>::schedule
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/progress/subgraph.rs:656:30
  14: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/progress/subgraph.rs:338:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/progress/subgraph.rs:304:17
  15: timely::worker::Wrapper::step::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/worker.rs:727:57
      core::option::Option<T>::map
             at usr/local/lib/rustlib/src/rust/library/core/src/option.rs:453:29
      timely::worker::Wrapper::step
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/worker.rs:727:26
  16: timely::worker::Worker<A>::step_or_park
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/worker.rs:375:38
  17: dataflow::server::Worker<A>::run
             at var/lib/buildkite-agent/builds/buildkite-builders-i-099ee983accf64f75-1/materialize/tests/src/dataflow/src/server.rs:436:13
  18: dataflow::server::serve::{{closure}}
             at var/lib/buildkite-agent/builds/buildkite-builders-i-099ee983accf64f75-1/materialize/tests/src/dataflow/src/server.rs:197:13
      timely::execute::execute::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/timely/src/execute.rs:284:22
      timely_communication::initialize::initialize_from::{{closure}}
             at cargo/git/checkouts/timely-dataflow-0f5e68a13a760402/26fd7cb/communication/src/initialize.rs:285:33
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at usr/local/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125:18
  19: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at usr/local/lib/rustlib/src/rust/library/std/src/thread/mod.rs:474:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at usr/local/lib/rustlib/src/rust/library/std/src/panic.rs:322:9
      std::panicking::try::do_call
             at usr/local/lib/rustlib/src/rust/library/std/src/panicking.rs:381:40
      std::panicking::try
             at usr/local/lib/rustlib/src/rust/library/std/src/panicking.rs:345:19
      std::panic::catch_unwind
             at usr/local/lib/rustlib/src/rust/library/std/src/panic.rs:396:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at usr/local/lib/rustlib/src/rust/library/std/src/thread/mod.rs:473:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at usr/local/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  20: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/boxed.rs:1307:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/boxed.rs:1307:9
      std::sys::unix::thread::Thread::new::thread_start
             at rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys/unix/thread.rs:71:17
  21: start_thread
  22: __clone
```

Reply:
It seems that this panic goes away if only one table is being sourced over the same publication.

Reply:
I found the issue, fixing it now

Reply:
@philip-stoev pushed a fix

Reply:
I am still getting this error even with aa0872dbf852d7d1a846ee84f45c8c97a37899c8. The full sequence of steps is:

```
postgres=# ALTER USER postgres WITH replication;
ALTER ROLE
postgres=# DROP PUBLICATION IF EXISTS mz_source;
NOTICE:  publication "mz_source" does not exist, skipping
DROP PUBLICATION
postgres=# DROP SCHEMA IF EXISTS public CASCADE;
NOTICE:  drop cascades to table t1
DROP SCHEMA
postgres=# CREATE SCHEMA public;
CREATE SCHEMA
postgres=# 
postgres=# CREATE TABLE t1 (f1 INTEGER);
CREATE TABLE
postgres=# INSERT INTO t1 VALUES (1);
INSERT 0 1
postgres=# ALTER TABLE t1 REPLICA IDENTITY FULL;
ALTER TABLE
postgres=# \q
pstoev@Ubuntu-2004-focal-64-minimal:~/petrosagg-pg-replication/test/pg-cdc$ psql -p 6875 -h 127.0.0.1 -U materialize
psql (12.5 (Ubuntu 12.5-0ubuntu0.20.04.1), server 9.5.0)
Type "help" for help.

materialize=> CREATE MATERIALIZED SOURCE "t1"
materialize->   FROM POSTGRES HOST 'host=127.0.0.1 port=5432 user=postgres password=postgres dbname=postgres'
materialize->   PUBLICATION 'mz_source' NAMESPACE 'public' TABLE 't1' (f1 INTEGER);
CREATE SOURCE
materialize=> select * from t1;
 f1 
----
  1
(1 row)

materialize=> \q
pstoev@Ubuntu-2004-focal-64-minimal:~/petrosagg-pg-replication/test/pg-cdc$ psql postgres
psql (12.5 (Ubuntu 12.5-0ubuntu0.20.04.1))
Type "help" for help.

postgres=# INSERT INTO t1 VALUES (1);
INSERT 0 1
postgres=# \q
pstoev@Ubuntu-2004-focal-64-minimal:~/petrosagg-pg-replication/test/pg-cdc$ psql -p 6875 -h 127.0.0.1 -U materialize
psql: error: could not connect to server: Connection refused
	Is the server running on host "127.0.0.1" and accepting
	TCP/IP connections on port 6875?
```

Reply:
This is because you never re-create the publication. I pushed a commit that prints out the error if it happens

Reply:
I got ```2021-02-23T07:28:05.617938Z ERROR panic: logical replication error: Error { kind: Closed, cause: None }``` once when messing with ```TAIL``` .( I think ```testdrive``` was constantly beating on the server while trying to run ```TAIL``` repeatedly) I assume this is expected at this stage. Let me know if you need a reproduction test case for it.